### DISCUSSION TITLE
Machine Learning in C

### DISCUSSION THREAD

**Discussion by eesuck1** [community/community#64870] [Programming Help]
Hello!
I recently started working on my Machine Learning project in pure C, for educational purposes, but as the project grows, the structure is getting more and more complicated
Therefore, I would be interested to hear your advice and possibly call on you to help in the development

https://github.com/eesuck1/NumC-CPlotLib

**Comment by Jaber-Saed** [comment]
Hello! It's great to hear that you're working on a Machine Learning project in C. Managing a growing project structure can indeed become challenging, but with good practices and organization, you can make the development process smoother. I've taken a look at your GitHub repository, and here are some suggestions and advice:

Modularization: As your project grows, consider breaking down your code into smaller, reusable modules. Each module can handle a specific functionality or a related set of functionalities. This promotes code reusability and makes it easier to manage and maintain.

Header Files and Separation of Concerns: Use header files to declare function prototypes, structures, and constants. This separation between implementation and declaration helps maintain clean code and avoids cyclic dependencies.

Documentation: Document your code using comments to explain the purpose of functions, structures, and any complex algorithms. Good documentation helps other developers (and your future self) understand your code without having to read through the entire implementation.

Version Control and Branching: Use version control (you're already using Git) to manage your project's history. Consider creating feature branches for new features or bug fixes. This allows you to work on isolated changes without affecting the main codebase until you're ready to merge.

Error Handling: Implement robust error handling mechanisms. When dealing with complex data and algorithms, errors can occur. Proper error handling can help you identify and resolve issues more efficiently.

Unit Testing: As your project grows, unit tests become invaluable. They ensure that individual components of your code are functioning correctly. This can help catch regressions and bugs early on.

Makefile or Build System: As your project becomes more complex, consider using a build system like Make to automate the compilation process. This helps ensure that changes are compiled consistently and can make deployment easier.

Code Reviews: If possible, collaborate with others who have C programming experience. Code reviews can provide valuable feedback, catch potential issues, and encourage best practices.

Memory Management: In C, memory management is crucial. Make sure to free memory that you allocate to avoid memory leaks.

Refactoring: Regularly review your codebase and refactor as needed. Refactoring can help improve the code's readability, maintainability, and performance.

Use Libraries: While you're building your project for educational purposes, don't hesitate to use existing libraries or frameworks for functionalities that are not the core focus of your project. For instance, you could use libraries for file I/O, memory management, or data manipulation.

Remember that building complex projects takes time, and it's okay to take iterative steps to improve your project's structure and maintainability. Don't hesitate to ask specific questions or seek assistance when facing challenges. If you encounter specific issues during your project's development, feel free to ask for guidance here. Good luck with your Machine Learning project in C!

**Reply by eesuck1** [reply]
Thank you for your reply!
But, as I understand it, most of the above has already been implemented in the project to some extent

I would be happy to hear any more advice

**Comment by pri-shav** [comment]
Hi everyone!
I'm impressed by your eagerness to use pure C in a machine learning project. It's fantastic that you are attempting something as difficult as this for educational purposes. Complexity frequently accompanies the evolution of a project, so feeling a little overwhelmed is normal.

You appear to be off to a good start based on what I could see in your GitHub repository [https://github.com/eesuck1/NumC-CPlotLib](https://github.com/eesuck1/NumC-CPlotLib). Taking into account the following tactics might be useful while handling increasing complexity:

Modularization: Divide your project into more manageable, smaller components. As a result, the overall structure may become easier to understand and manage.

Documentation: Ensure your code is well-documented. This not only helps others understand your code but will also be beneficial for you as the project expands. Tools like Doxygen can be handy.

Version Control: Since you're already using GitHub, make the most of version control. Regularly commit your changes and use branches to experiment with new features without affecting the main codebase.

Code Reviews: Encourage others to review your code. Fresh eyes can provide valuable insights and catch potential issues early on.

Testing: Implement testing early in your development process. This ensures that as you add new features, existing functionality remains intact.

Community Involvement: Seek advice and help from online communities or forums related to C and Machine Learning. The open-source community is often very supportive.

Ensure that your code has enough documentation. This will benefit you as the project progresses and make your code easier for other people to read. Additional resources, including oxygen, are also beneficial.

Version Control: Take use of version control as much as possible since you're already utilizing GitHub. Use branches to test new features without impacting the main core and commit your changes often.

Code Reviews: Ask other people to give your work a look. New eyes are able to see problems early and offer sage advice.

Testing: Include testing at the outset of the development process. This makes sure that functionality won't be lost when new features are implemented.
Source: [Machine Learning Course](https://www.igmguru.com/machine-learning-ai/machine-learning-certification-training/)

I hope this will help you.

**Comment by JHermosillaD** [comment]
Hi, sorry for being late. Currently, I use Python for machine learning, but I have worked a lot in robotics with limited hardware, so low-level languages were my bread and butter. Here are some of my tips:

1. I recommend you to switch to C++. Certainly, C cannot match Pandas or Numpy regarding dynamic structure handling. However, in C++ there are alternatives, such as Eigen or vector, that facilitate many implementations like ring buffers.
2. The more classical approaches, such as PCA, regression analysis, and the naive Bayes classifier, can be done with Eigen and the language's structures.
3. A natural step could be projects involving libraries that provide structures for training, and the king in that is OpenCV. It offers tutorials and an API to implement Haar cascade-based models such as HOG or Viola-Jones.
4. I have tried to get into ROS package development to use TensorFlow in a Google Coral with C++, but I don't recommend it. Even though it is "officially supported," in my experience, it is still too early to use it without compromising complications.
5. In my opinion, plotting something in C/C++ is unnecessary unless you are doing ray tracing or something resource-demanding. I would bind with another language for plotting or exporting the data.

I base this advice on my experience.
I hope you find them helpful.
